### ISSUE TITLE
Performance degradation

### ISSUE THREAD

**Issue by markgoodhead** [ugorji/go#218]
We just built one of our applications using current master (https://github.com/ugorji/go/commit/84cb69a8af8316eed8cf4a3c9368a56977850062) and found significant performance degradation in our RPC path (RPC round trip overhead from 5ms to ~1 second). Switching back to the commit we last built off fixes the issue (https://github.com/ugorji/go/commit/06e1e9d77f7fdab7f783b2e5e2936ea70c0c52df). This is on Ubuntu 14.04.5 LTS. 
CPU use is very low, but I noticed on perf top that there was a lot of this in the top 3:
[k] xen_hypercall_event_channel_op
Which doesn't appear at all using the older commit.
Let me know if there's anything else I can provide to help you debug.

**Comment by ugorji** [comment]
Can you provide more context
- pprof output when using old change
- pprof output when using new master
- code that you're running

Currently, we have optimized writing (so we buffer what we write before writing to socket at once) but not reading (so we don't read in more than necessary from the stream). If you want to optimize reading also, you should use the notes in the comments at https://godoc.org/github.com/ugorji/go/codec#pkg-variables ie

```
var handle codec.Handle     // codec handle
var conn io.ReadWriteCloser // connection got from a socket
var bufconn = struct {      // bufconn here is a buffered io.ReadWriteCloser
    io.Closer
    *bufio.Reader
    *bufio.Writer
}{conn, bufio.NewReader(conn), bufio.NewWriter(conn)}
var serverCodec = GoRpc.ServerCodec(bufconn, handle)
var clientCodec = GoRpc.ClientCodec(bufconn, handle)
```

**Comment by ugorji** [comment]
@markgoodhead response?

**Comment by markgoodhead** [comment]
Sorry @ugorji , we ended up just setting a go dep on the commit that works for us so I haven't gotten around to test out master since - will try your proposed read buffer solution and get back to you. In terms of pprof I did look at the CPU profile of the slow running program and it looked identical to the 'working' version. System CPU usage on the box was much higher than normal though.

**Comment by ugorji** [comment]
Thanks.

Please check it. After we spoke first, I added write buffering back as it has no downside, but kept read buffering off as it could allow us read more than we need, messing up something that needs to read right after us. E.g. imagine a chunk as json metadata, and next chunk as binary blob. With buffering, rpc could read into the binary blob, causing the next read to miss out on the beginning of the blob.

The slowdown is unfortunate, but better for users to be explicit in pre-creating the buffered reader so they can reuse it for subsequent reads.

**Comment by ugorji** [comment]
@markgoodhead 

Actually, your best way to do this, is to set the options for ReaderBufferSize and WriterBufferSize to be > 0. The codecs (when passed an io.Reader or io.Writer) will use buffers to read/write. 

```go
   var handle codec.JsonHandle // codec handle
   handle.ReaderBufferSize = 1024
   handle.WriterBufferSize = 1024
   var conn io.ReadWriteCloser // connection got from a socket
   var serverCodec = GoRpc.ServerCodec(conn, handle)
   var clientCodec = GoRpc.ClientCodec(conn, handle)
```

Please try this and let me know. I'm rounding out changes for a formal release, and would love to validate that this concern is mitigated easily with configuration option(s) on the Handle.

**Comment by ugorji** [comment]
This is bothering me. I remember that I explicitly turned on buffering on rpc for this reason.

I have expanded the options now:
- by default, we buffer on writes and reads
- a user can turn off rpc-attempt to buffer
- user can still pass a buffered connection or configure ReaderBufferSize or WriterBufferSize

This way, you are performant by default for the typical use-case, and we allow a power user to turn off implicit buffering.

Update coming soon.

**Comment by ugorji** [comment]
Closed via 14c9063
